### DISCUSSION TITLE
Unable to import git module in python

### DISCUSSION THREAD

**Discussion by ANIKET200915** [gitpython-developers/GitPython#1841] [Q&A] [answered]
Today while working on a project on Python I needed to import git module so in I opened CMD and entered the following command `pip install GitPython` to install git then when I went to test the module on CMD by opening python shell and typing `import git` I got an error message 

```
`ImportError: Bad git executable.
The git executable must be specified in one of the following ways:
    - be included in your $PATH
    - be set via $GIT_PYTHON_GIT_EXECUTABLE
    - explicitly set via git.refresh()

All git commands will error until this is rectified.

This initial message can be silenced or aggravated in the future by setting the
$GIT_PYTHON_REFRESH environment variable. Use one of the following values:
    - quiet|q|silence|s|silent|none|n|0: for no message or exception
    - warn|w|warning|log|l|1: for a warning message (logged at level CRITICAL, displayed by default)
    - error|e|exception|raise|r|2: for a raised exception

Example:
    export GIT_PYTHON_REFRESH=quiet


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310\lib\site-packages\git\__init__.py", line 142, in <module>
    raise ImportError("Failed to initialize: {0}".format(_exc)) from _exc
ImportError: Failed to initialize: Bad git executable.
The git executable must be specified in one of the following ways:
    - be included in your $PATH
    - be set via $GIT_PYTHON_GIT_EXECUTABLE
    - explicitly set via git.refresh()

All git commands will error until this is rectified.

This initial message can be silenced or aggravated in the future by setting the
$GIT_PYTHON_REFRESH environment variable. Use one of the following values:
    - quiet|q|silence|s|silent|none|n|0: for no message or exception
    - warn|w|warning|log|l|1: for a warning message (logged at level CRITICAL, displayed by default)
    - error|e|exception|raise|r|2: for a raised exception

Example:
    export GIT_PYTHON_REFRESH=quiet`
```

**Comment by EliahKagan** [comment, marked as answer]
> I opened CMD and entered the following command `pip install GitPython` to install git

GitPython provides a `git` module for Python code, but it does not provide [Git](https://git-scm.com/) itself. GitPython does most of its work by using the external `git` command that [Git](https://git-scm.com/) provides. So GitPython needs [Git](https://git-scm.com/) to be present.

For the GitPython `git` module to make use of this external `git` command, the external command must either be runnable as `git` from the environment in which you run the Python process that uses GitPython, or specially configured as described in the message you saw. In most cases, it either is present as `git` or is intended to be present as `git`.

Is [Git](https://git-scm.com/) installed? If it is installed and usable, then running `git version` from the command line should show something like `git version 2.43.0.windows.1`.

If you are able to successfully run `git version` from the same command line that you run `python` on, then GitPython should usually be able to use it as well, and if it cannot then this might be due to a bug in GitPython. But if `git version` does not work when you run it, then GitPython will usually be unable to run the external `git` command for the same reason you are unable to run it directly (whatever that reason happens to be).
